code/modules/projectiles/guns/specialist.dm

```diff
@@ -1296,3 +1296,95 @@
 	. = ..()
 	if(last_signal_flare_name)
 		. += SPAN_NOTICE("The last signal flare fired has the designation: [last_signal_flare_name]")
+
+//-------------------------------------------------------
+//P9 Sonic Harpoon Artillery Remote Projectile(SHARP) Rifle
+
+/obj/item/weapon/gun/rifle/sharp
+	name = "\improper P9 SHARP Rifle"
+	desc = "A experimental harpoon launcher rifle manufactured by Armat Systems. It's specialized for specific ammo types out of a 10-round magainze, best used for area denial and disruption."
+	icon_state = "sharprifle"
+	item_state = "sharp"
+	fire_sound = 'sound/weapons/gun_sharp.ogg' // to fix
+	reload_sound = 'sound/weapons/handling/m41_reload.ogg' // to fix
+	unload_sound = 'sound/weapons/handling/m41_unload.ogg' // to fix
+	unacidable = TRUE
+	indestructible = 1
+	muzzle_flash = null
+
+	current_mag = /obj/item/ammo_magazine/rifle/sharp/explosive
+	attachable_allowed = list(/obj/item/attachable/magnetic_harness)
+	matter = list("metal" = 6000)
+
+	aim_slowdown = SLOWDOWN_ADS_SPECIALIST
+	wield_delay = WIELD_DELAY_NORMAL
+	flags_gun_features = GUN_SPECIALIST|GUN_WIELDED_FIRING_ONLY|GUN_CAN_POINTBLANK|GUN_AMMO_COUNTER
+
+	flags_item = TWOHANDED|NO_CRYO_STORE
+	map_specific_decoration = TRUE
+
+	var/explosion_delay_sharp = FALSE
+	var/list/sharp_tracked_mob_list = list()
+
+/obj/item/weapon/gun/rifle/sharp/set_gun_attachment_offsets()
+	attachable_offset = list("muzzle_x" = 32, "muzzle_y" = 17,"rail_x" = 12, "rail_y" = 20, "under_x" = 23, "under_y" = 13, "stock_x" = 24, "stock_y" = 13)
+
+/obj/item/weapon/gun/rifle/sharp/set_gun_config_values()
+	..()
+	fire_delay = FIRE_DELAY_TIER_1
+	accuracy_mult = BASE_ACCURACY_MULT
+	scatter = SCATTER_AMOUNT_NONE
+	damage_mult = BASE_BULLET_DAMAGE_MULT
+	recoil = RECOIL_OFF
+	burst_amount = 0 //just incase
+
+/obj/item/weapon/gun/rifle/sharp/unique_action(mob/user)
+	track(user)
+
+/obj/item/weapon/gun/rifle/sharp/proc/track(mob/user)
+	var/mob/living/carbon/human/M = user
+
+	var/max_count = 5 //max number of tracking
+	var/target
+	var/direction = -1
+	var/atom/areaLoc = null
+	var/output = FALSE
+
+	var/x = sharp_tracked_mob_list.len - max_count
+	for(var/i=0,i<x,++i)
+		popleft(sharp_tracked_mob_list)
+
+	for(var/mob/living/mob_tracked as mob in sharp_tracked_mob_list)
+		if(!QDELETED(mob_tracked))
+			if(M.z == mob_tracked.z)
+				var/dist = get_dist(M,mob_tracked)
+				target = dist
+				direction = get_dir(M,mob_tracked)
+				areaLoc = loc
+
+			if(target < 900)
+				output = TRUE
+				var/areaName = get_area_name(areaLoc)
+				to_chat(M, SPAN_NOTICE("\The [mob_tracked] is [target > 10 ? "approximately <b>[round(target, 10)]</b>" : "<b>[target]</b>"] paces <b>[dir2text(direction)]</b> in <b>[areaName]</b>."))
+	if(!output)
+		to_chat(M, SPAN_NOTICE("There is nothing currently tracked."))
+
+	return
```

Comment:
I could do something REALLY cool with this using TG's nuke pinpointer TGUI

we'll talk about it later